---
title: "Exercise 1"
format: html
editor: source
editor_options: 
  chunk_output_type: inline
---

## Learning Objectives

By the end of this exercise, you will be able to:

-   Understand the structure of a `SpatialExperiment` object.
-   Access and interpret `spatialCoords` and `imgData`.
-   Perform basic subsetting operations on `SpatialExperiment` objects.
-   Combine multiple `SpatialExperiment` objects.
-   Apply image transformations (rotation, mirroring) to spatial data.

## Libraries

```{r}
#| echo: false
#| message: false
#| warning: false
#| output: false

# load libraries to automatically include in renv
library(downlit)
library(xml2)
library(GenomeInfoDb)

```

```{r}
#| message: false
#| warning: false
#| output: false

# Load necessary R libraries for spatial transcriptomics data analysis.
# library(DropletUtils)
# library(BumpyMatrix)
library(SpatialExperiment)
library(VisiumIO)
library(arrow)
library(qs2)
library(ggspavis)
library(patchwork)
library(dplyr)
library(arrow)
library(scuttle)
```

## Data for the course

We will start with downloading the data. We will work on a VisiumHD dataset from a human colon cancer study [published recently](https://www.nature.com/articles/s41588-025-02193-3). A Visium HD slide from one patient (P2CRC) used in the article is available [from the 10X website](https://www.10xgenomics.com/datasets/visium-hd-cytassist-gene-expression-libraries-of-human-crc). We will focus on the "binned output" provided available as an output of the Space Ranger pipeline. 
The dataset contains normal adjacent tissue (NAT) and colorectal carcinoma (CRC) from 5 patients.  
We will start downloading and processing one of the Human colon cancer samples (P1).   

> We need more explanation here!
Julien: Maybe we could ask questions on the dataset that they can find out in the paper or 10X website, such as: how was the tissue processed for spatial profiling? How were the RNA molecules captured? Which version of the SpaceRanger tool was used to process the data? What is the "binned output" about? They can have a look at this page https://www.10xgenomics.com/support/software/space-ranger/latest/analysis/outputs/output-overview.


```{r}

if (!dir.exists("data/Human_Colon_Cancer_P1/binned_outputs/")) {
  dir.create("data")
  download.file(
    url = "https://seq-spatial-transcriptomics-training.s3.eu-central-1.amazonaws.com/Human_Colon_Cancer_P1.tar.gz",
    destfile = "data/Human_Colon_Cancer_P1.tar.gz"
  )

  untar(
    tarfile = "data/Human_Colon_Cancer_P1.tar.gz", exdir = "data/"
  )

  file.remove("data/Human_Colon_Cancer_P1.tar.gz")
} else {
  message("Data exists, please proceed to next steps!")
}
```

We will import the dataset into a `SpatialExperiment` object. For these exercises, we choose to use the largest bin size of 16 µm, and we will subset the dataset for faster processing:

```{r}
# Import Visium HD data from Space Ranger output into a SpatialExperiment object.
spe <- TENxVisiumHD(
  spacerangerOut = "data/Human_Colon_Cancer_P1/",
  processing = "filtered",
  format = "h5",
  images = "lowres",
  bin_size = "016"
) |>
  import()

# we will subset the dataset for computational and visual reasons
spe <- spe[, spatialCoords(spe)[, 1] * scaleFactors(spe) > 70 &
               spatialCoords(spe)[, 1] * scaleFactors(spe) < 130 &
               spatialCoords(spe)[, 2] * scaleFactors(spe) > 200 &
               spatialCoords(spe)[, 2] * scaleFactors(spe) < 260]
```





::: callout-important
## Exercise

What other bin size do we have available?
:::

::: {.callout-tip collapse="true"}
## Answer

In `data/Human_Colon_Cancer_P1/binned_outputs` you will find three directories:

```         
square_002um
square_008um
square_016um
```

So in addition to 16 µm bins, we also have 2 and 8 µm bins available. The smallest capture bin is 2µm^2, see Fig. 2a in the paper 
:::

## Exploring the object

In this first exercise, we will dive into the `SpatialExperiment` object, a cornerstone of spatial transcriptomics analysis in `R`. We will explore the different components of the loaded object, and perform basic manipulations. By the end of this session, you will have a solid understanding of how to handle spatial transcriptomics data within the `R`/`Bioconductor` ecosystem.

![Overview of the `SpatialExperiment` class structure](../assets/images/spe.png)

The different slots of the `SpatialExperiment` object can be approached with several accessor functions that correspond to their names. As the `SpatialExperiment` is an extension of the `SingleCellExperiment` class, many of the functions are shared. 

::: callout-important
## Exercise

Check out the outputs of the following functions:

-   `colData()`
-   `spatialCoords()`
-   `rowData()`
-   `assay()`
-   `imgRaster()`
-   `reducedDims()`

Questions:

-   What kind of data is in the different slots? Which slots are specific to the `SpatialExperiment` class compared to the `SingleCellExperiment` class?
-   How many spots and genes do we have retained in the `spe` object?
-   Are spots in the tissue?
:::

::: {.callout-tip collapse="true"}
## Answer

The slot `colData` contains metadata of each of the 14207 spots:

```{r}
colData(spe) |> head()
ncol(spe)
```

We also have the column `in_tissue` here, and we can check whether all the spots are in the tissue:

```{r}
colData(spe) |> as.data.frame() |>
  group_by(in_tissue) |> summarise(number = n())
```

The slot `spatialCoords` maps each spot to the full resolution image:

```{r}
spatialCoords(spe) |> head()
```

The slot `rowData` contains metadata about the 18085 genes for which we measured expression:

```{r}
rowData(spe) |> head()
nrow(spe)
```

The assay is the core matrix of the object. Corresponding to the genes (rows) and spots (columns):

```{r}
assay(spe)
```
Of course, the H&E image is also included:

```{r}
imgRaster(spe) |> plot()
```

The slot for `reducedDims` is empty for now, as we didn't perform any dimensionality reduction calculations yet:

```{r}
reducedDims(spe)
```
:::

We saw that the rownames of the object were Ensembl IDs. However, it's more convenient to have gene symbols. Therefore, we replace the ensembl IDs, with the symbols stored in `rowData`:

```{r}
rownames(spe) <- uniquifyFeatureNames(ID=rowData(spe)$ID, names=rowData(spe)$Symbol)
```

To get a visual overview the slide we can use `ggspavis::plotVisium`. First without spots:

```{r}
plotVisium(spe, spots = FALSE)
```

And with spots:

```{r}
plotVisium(spe, point_shape = 22, point_size = 0.5) 
```

Question: Why do we see only spots in a specific area?

::: {.callout-tip collapse="true"}
## Answer
After loading the object, we have choosen tosubset the dataset for faster processing. Therefor we have spots only for the certain area selected. You can also visualise only the selected area, see next exercise.

:::

Coloured by e.g., gene expression levels of a selected gene (ie. PIGR): 

```{r}
plotVisium(spe, 
           annotate = "PIGR", 
           assay = "counts", 
           zoom = TRUE, 
           point_size = 1,
           point_shape = 22)
```
::: callout-important
## Exercise

Check out the usage of `plotVisium` with `?plotVisium`. Create a plot that zooms in into the spots, and colors the spots according to column `array_row`.

In which slot is `array_row` stored and what does it represent?
:::

::: {.callout-tip collapse="true"}
## Answer

The column `array_row` is stored in `colData`, it represents in which row the spot is positioned. Of course, plotting this value is not particularly useful for biology. We will visualize more sensible values in the next chapter, when we do the quality control.

We can use the argument `annotate` to color according to a column in `colData`, and we can zoom in by specifying `zoom = TRUE`:

```{r}
p <- plotVisium(spe, annotate = "array_row", zoom = TRUE, point_shape = 22)
p
```
:::



## retain only observations that map to tissue 

Subset the SpatialExperiment object to retain only spots that are within the tissue area

sub <- spe[, colData(spe)$in_tissue]

How many spots were present in "spe" object? And how many are left after selecting for in-tissue spots only (object sub)?

::: {.callout-tip collapse="true"}
## Answer

```{r}
dim(spe)
dim(sub)
```

Both objects have the same dimensions. Our spe object is a subset of tissue slice were all spots map tissue. Therefore the filtering here is not relevant. 

:::


# Save the object

```{r}
dir.create("results/day1", showWarnings = FALSE)
#gc()
#qs_save(spe, file = "results/day1/01_spe.qs2")
saveHDF5SummarizedExperiment(spe, dir="results/", prefix="01_spe", replace=FALSE,
                                chunkdim=NULL, level=NULL, as.sparse=NA,
                                verbose=NA)
```

Clear your environment

```{r}
#| eval: false
rm(list = ls())
gc()
.rs.restartR()
```

::: callout-important
**Key Takeaways:**

-   The `SpatialExperiment` class is a versatile container for spatial transcriptomics data.
-   It allows for easy access and manipulation of spatial coordinates, image data, and assay data.
-   Various operations like subsetting, combining, and image transformations can be performed efficiently.
:::
