---
title: "Exercise 1"
format: html
editor: source
editor_options: 
  chunk_output_type: console
execute:
  warning: false
---

## Learning Objectives

By the end of this exercise, you will be able to:

-   Understand the structure of a `SpatialExperiment` object.
-   Access and interpret `spatialCoords` and `imgData`.
-   Perform basic subsetting operations on `SpatialExperiment` objects.
-   Combine multiple `SpatialExperiment` objects.
-   Apply image transformations (rotation, mirroring) to spatial data.

## Libraries

```{r day1-1-SpatialExperiment-1 }
#| echo: false
#| message: false
#| warning: false
#| output: false

# load libraries to automatically include in renv
library(downlit)
library(xml2)
library(GenomeInfoDb)
library(scrapper)
library(SingleR)
library(BayesSpace)
library(xgboost)
library(namer)
library(styler)
```

```{r day1-1-SpatialExperiment-2 }
#| message: false
#| warning: false
#| output: false

# Load necessary R libraries for spatial transcriptomics data analysis.
library(SpatialExperiment)
library(VisiumIO)
library(arrow)
library(qs2)
library(ggspavis)
library(patchwork)
library(dplyr)
library(arrow)
library(scuttle)
library(HDF5Array)
```

## Data for the course

We will start with downloading the data. We will work on a VisiumHD dataset from a human colon cancer study [published recently](https://www.nature.com/articles/s41588-025-02193-3). A Visium HD slide from one patient (P2CRC) used in the article is available [from the 10X website](https://www.10xgenomics.com/datasets/visium-hd-cytassist-gene-expression-libraries-of-human-crc). We will focus on the "binned output" provided available as an output of the Space Ranger pipeline. 

The dataset contains normal adjacent tissue (NAT) and colorectal carcinoma (CRC) from 5 patients.  
We will start downloading and processing one of the Human colon cancer samples (P1).   

::: callout-important
## Exercise 1a

Have a look at the paper and 10X website, to find out how was the tissue processed for spatial profiling. How are the transcript molecules measured? 

Which version of the Space Ranger tool was used to process the data, and what is the output folders containing? This page will be useful: https://www.10xgenomics.com/support/software/space-ranger/latest/analysis/outputs/output-overview. Additionally the Space Ranger "web summary" gives a good idea about the quality of the processed sample: https://cf.10xgenomics.com/samples/spatial-exp/3.0.0/Visium_HD_Human_Colon_Cancer/Visium_HD_Human_Colon_Cancer_web_summary.html
:::

::: {.callout-tip collapse="true"}
## Answer
The measurements of RNA molecules is indirect, through capture of ligated probe pairs targeting the whole protein-coding transcriptome.

An instrument called Visium CytAssist is used to control reagent flow and better capture target probes from the tissue (in particular to limit problematic free diffusion of transcripts).

The dataset was processed with Space Ranger version 3.0.0. Note that this is not the latest version: v4.0 is the latest and it has cool feature, such as a segmented output.
:::

```{r day1-1-SpatialExperiment-3 }
if (!dir.exists("data/Human_Colon_Cancer_P1/binned_outputs/")) {
  dir.create("data")
  download.file(
    url = "https://seq-spatial-transcriptomics-training.s3.eu-central-1.amazonaws.com/Human_Colon_Cancer_P1.tar.gz",
    destfile = "data/Human_Colon_Cancer_P1.tar.gz"
  )

  untar(
    tarfile = "data/Human_Colon_Cancer_P1.tar.gz", exdir = "data/"
  )

  file.remove("data/Human_Colon_Cancer_P1.tar.gz")
} else {
  message("Data exists, please proceed to next steps!")
}
```

We will import the dataset into a `SpatialExperiment` object. For these exercises, we choose to use the largest bin size of 16 µm, and we will subset the dataset for faster processing during the practicals:

```{r day1-1-SpatialExperiment-4 }
# Import Visium HD data from Space Ranger output into a SpatialExperiment object.
spe <- TENxVisiumHD(
  spacerangerOut = "data/Human_Colon_Cancer_P1/",
  processing = "filtered",
  format = "h5",
  images = "lowres",
  bin_size = "016"
) |>
  import()

# we subset the dataset based on x and y coordinates of the tissue slice
spe <- spe[, spatialCoords(spe)[, 1] * scaleFactors(spe) > 70 &
  spatialCoords(spe)[, 1] * scaleFactors(spe) < 130 &
  spatialCoords(spe)[, 2] * scaleFactors(spe) > 200 &
  spatialCoords(spe)[, 2] * scaleFactors(spe) < 260]
```


::: callout-important
## Exercise 1b

**What other bin sizes are available? Which is the smallest capture unit and what do the other bin sizes mean? Why wouldn't you always want to work with the smallest available bin?**

:::

::: {.callout-tip collapse="true"}
## Answer

In `data/Human_Colon_Cancer_P1/binned_outputs` you will find three directories:

```         
square_002um
square_008um
square_016um
```

The capture bins are 2µm^2 (adjacent) on the VisiumHD slide, see Fig. 2a in the paper. Larger bins are pools of 2 µm bins (16 bins for an 8 µm bin, 64 bins for a 16 µm bin). Working with those could be handy because the number of molecules captured will be higher, reducing the sparsity of the data at the cost of resolution.
:::

## Exploring the object

In this first exercise, we will dive into the `SpatialExperiment` object, a cornerstone of spatial transcriptomics analysis in `R`. We will explore the different components of the loaded object, and perform basic manipulations. By the end of this session, you will have a solid understanding of how to handle spatial transcriptomics data within the `R`/`Bioconductor` ecosystem.

![Overview of the `SpatialExperiment` class structure](../assets/images/spe.png)

The different slots of the `SpatialExperiment` object can be approached with several accessor functions that correspond to their names. As the `SpatialExperiment` is an extension of the `SingleCellExperiment` class, methods developed to be applied on `SingleCellExperiment` objects (or by extension on `SummarizedExperiment` objects) can also be applied.

::: callout-important
## Exercise 2

Check out the outputs of the following functions, that you should already be familiar with:

-   `colData()`
-   `rowData()`
-   `assay()`
-   `reducedDims()`

Check out the outputs of the following functions, which are specific to the `SpatialExperiment` class:

-   `spatialCoords()`
-   `imgData()`
-   `imgRaster()`

Questions:

-   What kind of data is in the different slots?
-   How many spots and genes do we have retained in the `spe` object?
-   Are all spots within the tissue area?
:::

::: {.callout-tip collapse="true"}
## Answer

The slot `colData` contains metadata of each of the 14207 spots:

```{r day1-1-SpatialExperiment-5 }
colData(spe) |> head()
ncol(spe)
```

We have a column `in_tissue` here, and we can check whether spots are covered by the tissue:

```{r day1-1-SpatialExperiment-6 }
colData(spe) |>
  as.data.frame() |>
  group_by(in_tissue) |>
  summarise(number = n())
```
The slot `rowData` contains metadata about the 18085 genes for which we measured expression:

```{r day1-1-SpatialExperiment-7 }
rowData(spe) |> head()
nrow(spe)
```

The assay is the core count matrix of the object, corresponding to all genes (rows) and all spots (columns):

```{r day1-1-SpatialExperiment-8 }
assay(spe)
```
The slot for `reducedDims` is empty for now, as we didn't perform any dimensionality reduction calculations yet:

```{r day1-1-SpatialExperiment-9 }
reducedDims(spe)
```

The slot `spatialCoords` maps each spot to the full resolution image:

```{r day1-1-SpatialExperiment-10 }
spatialCoords(spe) |> head()
```
The slot `imgData` lists all available images linked to the dataset, their resolution and a scaling factor used to convert pixels coordinates relative to the full resolution image coordinates (above). 

```{r day1-1-SpatialExperiment-11 }
imgData(spe)
```

The H&E image can be exported asa raster image for plotting using:

```{r day1-1-SpatialExperiment-12 }
imgRaster(spe) |> plot()
```

:::

We saw that the rownames of the object were Ensembl IDs. However, it's more convenient to have gene symbols. Therefore, we replace the Ensembl IDs, with the symbols stored in `rowData` (making sure to keep the ensembl ID if these are not unique):

```{r day1-1-SpatialExperiment-13 }
rownames(spe) <- uniquifyFeatureNames(ID = rowData(spe)$ID, names = rowData(spe)$Symbol)
```

To get a visual overview the slide we can use `ggspavis::plotVisium`. First without spots:

```{r day1-1-SpatialExperiment-14 }
plotVisium(spe, spots = FALSE)
```

And with spots overlayed:

```{r day1-1-SpatialExperiment-15 }
plotVisium(spe, point_shape = 22, point_size = 0.5)
```

::: callout-important
## Exercise 3

Why do we see only spots in a specific area?
:::

::: {.callout-tip collapse="true"}
## Answer
After loading the object above, we subsetted the dataset for faster processing. Therefore we only see the spots overlayed with the selected area. You can also visualize only the selected area with the `zoom` option, see next exercise.

:::

It's possible to colour spots according to some variable, e.g., the number of read counts for a selected gene (i.e., PIGR): 

```{r day1-1-SpatialExperiment-16 }
plotVisium(spe,
  annotate = "PIGR",
  assay = "counts",
  zoom = TRUE,
  point_size = 1,
  point_shape = 22
)
```
::: callout-important
## Exercise 4

Check out the usage of `plotVisium` with `?plotVisium`. Create a plot that zooms into the spots, and colors the spots according to column `array_row`.

In which slot is `array_row` stored and what does it represent?
:::

::: {.callout-tip collapse="true"}
## Answer

The column `array_row` is stored in `colData`, it represents in which row the spot is positioned. Of course, plotting this value is not particularly useful for biology. We will visualize more sensible values in the next chapter, when we do the quality control.

We use the argument `annotate` to color according to a column in `colData`, and we zoom in by specifying `zoom = TRUE`:

```{r day1-1-SpatialExperiment-17 }
p <- plotVisium(spe, annotate = "array_row", zoom = TRUE, point_shape = 22)
p
```
:::

## Retain only observations that map to tissue 

::: callout-important
## Exercise 5

Subset the SpatialExperiment object (and save it into a new object) to retain only spots that are within the tissue area

How many spots were present in "spe" object? And how many are left after selecting for in-tissue spots only (object sub)?
:::

::: {.callout-tip collapse="true"}
## Answer

```{r day1-1-SpatialExperiment-18 }
# Subset to in-tissue spots only
sub <- spe[, colData(spe)$in_tissue]
```

```{r day1-1-SpatialExperiment-19 }
# Check dimensions of original and subsetted objects
dim(spe)
dim(sub)
```

Both objects have the same dimensions. Our spe object is a subset of tissue slice were all spots map tissue. Therefore the filtering here is not relevant. 
:::


## Save the object

Save the filtered `SpatialExperiment` object for the next steps:
```{r day1-1-SpatialExperiment-20 }
dir.create("results/day1", showWarnings = FALSE, recursive = TRUE)
saveHDF5SummarizedExperiment(spe,
  dir = "results/day1", prefix = "01.1_spe", replace = TRUE,
  chunkdim = NULL, level = NULL, as.sparse = NA,
  verbose = NA
)
```

Clear your environment

```{r day1-1-SpatialExperiment-21 }
#| eval: false
rm(list = ls())
gc()
.rs.restartR()
```

::: callout-important
**Key Takeaways:**

-   The `SpatialExperiment` class is a versatile container for spatial transcriptomics data.
-   It allows for easy access and manipulation of spatial coordinates, image data, and assay data.
-   Various operations like subsetting, combining, and image transformations can be performed efficiently.
:::
