---
title: "Exercise 4"
format: html
editor: source
editor_options: 
  chunk_output_type: console
---


## Clustering and Spatially-aware clustering

In the next section we will explore clustering methods for spatial transcriptomics data, including both non-spatially aware and spatially-aware approaches. 

## Learning Objectives
By the end of this exercise, you will be able to:

-  Perform non-spatially aware clustering using graph-based methods (Leiden clustering).
-  Perform spatially-aware clustering using Bayesian modeling (BayesSpace) and graph-based methods (Banksy + Leiden).
-  Visualize and compare clustering results from different methods on tissue slides.

## Libraries
```{r}
#| message: false
#| warning: false
#| output: false

library(Banksy)
library(igraph)
library(ggspavis)
library(BayesSpace)
library(nnSVG)
library(HDF5Array) # For loading object
library(patchwork)
library(SpatialExperiment)
library(scran) 
library(scater)
```



## Clustering

Load the previously saved `SpatialExperiment` object which contains information on the HVGs and dimensionationality reduction computed with Banksy and PCA.

```{r}
#spe.hvg <- loadHDF5SummarizedExperiment(dir="results/day1", prefix="01.3_spe_tmp_tmp")
spe <- loadHDF5SummarizedExperiment(dir="results/day1", prefix="01.3_spe")
```


### Non-spatial aware clustering

Clustering methods can also be categorized into non-spatially aware and spatially-aware methods. We will first explore non-spatially aware clustering using graph-based methods.

#### Leiden clustering

As commonly done in single-cell RNA-seq analysis, we can perform graph-based clustering using the Leiden algorithm on a shared nearest-neighbor (SNN) graph constructed from the PCA results.

```{r}
# build cellular shared nearest-neighbor (SNN) graph
g <- buildSNNGraph(spe, use.dimred="PCA", type="jaccard", k=20)
# cluster using Leiden community detection algorithm
k <- cluster_leiden(g, objective_function="modularity", resolution=0.6)
# assign cluster labels to spe object
table(spe$Leiden <- factor(k$membership))
```


### Spatialy aware clustering

Clustering methods can also incorporate spatial information to identify spatial domains or regions in the tissue. There are different kinds of methods, here we will explore two different approaches: Bayesian modeling (probabilistic) and spatially-aware graph-based clustering.

#### Probabilistic: BayesSpace

We can use the `BayesSpace` package to perform spatially aware clustering using a Bayesian modeling approach.

```{r}
# prepare data for 'BayesSpace'
# skipping PCA (already computed)
spe <- spatialPreprocess(spe, skip.PCA=TRUE)
# perform spatial clustering with 'BayesSpace'
# using 'd=10' PCs and targeting 'q=8' clusters
spe <- spatialCluster(spe, q=8, d=10, nrep=1e3, burn.in=100) 
table(spe$BayesSpace <- factor(spe$spatial.cluster))
```


#### Graph-based: Banksy

We have computed a principal components using Banksy mathod already in the previous step. We will now used them (use Banksy dimentionality reduction space) to perdoem SNN graph-based Leiden clustering based on them, in order to obtain spatially-aware clusters.
When building the SNN graph, we will use the same parameters we have used before for the non-spatially aware graph, but using banksy PCs.

```{r}
# perform SNN graph-based clustering on 'Banksy' PCs using
g <- buildSNNGraph(spe, use.dimred="PCA_banksy", type="jaccard", k=20)
# cluster using Leiden community detection algorithm
k <- cluster_leiden(g, objective_function="modularity", resolution=0.6)
table(spe$Banksy <- factor(k$membership))
```


## Visualisation of the different clustering methods:

Let's visualize the clustering results from the different methods on the tissue slide.

```{r}
# Define the three clustering methods to compare
ks <- c("BayesSpace", "Leiden", "Banksy")#, "CellCharter") 
# Plot spatial distribution of clusters for each method
plots <- c(lapply(ks, \(.) {
    plt <- plotVisium(spe, annotate=., zoom = T)
    plt$layers[[1]]$aes_params$stroke <- 0
    plt$layers[[1]]$aes_params$size <- 0.2
    plt
}) ,  
  plotVisium(spe, zoom = T,  spots=FALSE )) # add a plot of the tissue without spots
# combine plots
plots |>  wrap_plots(nrow=2) &
    #scale_color_manual(values=unname(pals::trubetskoy())) &
    theme(legend.key.size=unit(0, "lines"), legend.justification="left")
```

It can help us to visualise the clusters one by one, in order to better see their spatial distribution. We take as an example the Banksy clusters:

```{r}
# plot selected clusters in order of frequency,
# highlighting cells assigned to cluster 'k'
lapply(tail(names(sort(table(spe$Banksy))), 8), \(k) {
    spe$foo <- spe$Banksy == k
    spe <- spe[, order(spe$foo)]
    plt <- plotCoords(spe, annotate="foo")
    plt$layers[[1]]$aes_params$stroke <- 0
    plt$layers[[1]]$aes_params$size <- 0.2
    plt + ggtitle(k)
}) |>
    wrap_plots(nrow=3) &
    scale_color_manual(values=c("lavender", "purple")) &
    theme(plot.title=element_text(hjust=0.5), legend.position="none")
```


::: callout-important
## Bonus question
Visualise the clusters obtained with BayesSpace and Leiden methods one by one, and compare their spatial distribution to the Banksy clusters. Do you observe any similarities or differences in the spatial patterns?
:::   


## Save the object

Save the `SpatialExperiment` object for the next steps:

```{r}
saveHDF5SummarizedExperiment(spe, dir="results/day1", prefix="01.4_spe", replace=TRUE,
                                chunkdim=NULL, level=NULL, as.sparse=NA,
                                verbose=NA)
```

Clear your environment:
```{r}
#| eval: false
rm(list = ls())
gc()
.rs.restartR()
```